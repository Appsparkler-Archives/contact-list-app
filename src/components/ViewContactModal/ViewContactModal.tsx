import * as React from "react";
import Button from "@mui/material/Button";
import Dialog from "@mui/material/Dialog";
import AppBar from "@mui/material/AppBar";
import Toolbar from "@mui/material/Toolbar";
import IconButton from "@mui/material/IconButton";
import Typography from "@mui/material/Typography";
import CloseIcon from "@mui/icons-material/Close";
import { FullScreenModalTransition } from "components/FullScreenModalTransition/FullScreenModalTransition";
import { Avatar, Box } from "@mui/material";
import {
  Business as BusinessIcon,
  Category as CategoryIcon,
  ChromeReaderMode as ChromeReaderModeIcon,
  Email as EmailIcon,
  Female as FemaleIcon,
  House as HouseIcon,
  Male as MaleIcon,
  PhoneAndroid as PhoneAndroidIcon,
} from "@mui/icons-material";
import { TContactInfoItem, TViewContactModalFC } from "types";
import { blue, pink } from "@mui/material/colors";

export const ViewContactModal: TViewContactModalFC = ({ contact }) => {
  const [open, setOpen] = React.useState(false);

  const handleClickOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  return (
    <Box>
      <IconButton aria-label="read" color="primary" onClick={handleClickOpen}>
        <ChromeReaderModeIcon />
      </IconButton>
      <Dialog
        fullScreen
        open={open}
        onClose={handleClose}
        TransitionComponent={FullScreenModalTransition}
      >
        <AppBar sx={{ position: "relative" }}>
          <Toolbar>
            <IconButton
              edge="start"
              color="inherit"
              onClick={handleClose}
              aria-label="close"
            >
              <CloseIcon />
            </IconButton>
            <Typography sx={{ ml: 2, flex: 1 }} variant="h6" component="div">
              View Contact
            </Typography>
            <Button autoFocus color="inherit" onClick={handleClose}>
              close
            </Button>
          </Toolbar>
        </AppBar>
        <Box p={2} display={"flex"} flexDirection={"column"} gap={2}>
          <Avatar
            sx={{ bgcolor: contact.gender === "male" ? blue[500] : pink[500] }}
          >
            {contact.gender === "male" ? (
              <MaleIcon fontSize={"large"} />
            ) : (
              <FemaleIcon fontSize={"large"} />
            )}
          </Avatar>
          <Typography variant="h5"> {contact.name}</Typography>
          <InfoItem icon={<PhoneAndroidIcon />} info={contact.mobile} />
          <InfoItem icon={<EmailIcon />} info={contact.email} />
          <InfoItem icon={<HouseIcon />} info={contact.address} />
          <InfoItem icon={<BusinessIcon />} info={contact.company} />
          <InfoItem icon={<CategoryIcon />} info={contact.type} />
        </Box>
      </Dialog>
    </Box>
  );
};

const InfoItem: TContactInfoItem = ({ icon, info }) => {
  if (!info) return null;
  return (
    <Box display="flex" flexDirection={"row"} alignItems={"center"} gap={2}>
      {icon}
      <Typography variant="h6">{info}</Typography>
    </Box>
  );
};
