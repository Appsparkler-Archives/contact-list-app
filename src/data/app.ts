import { IContactFormData } from "types";

export type TAppFC = React.FC<{
  contacts: IContactFormData[];
}>;
