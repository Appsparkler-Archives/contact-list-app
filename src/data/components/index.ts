export * from "./standard-select-field";
