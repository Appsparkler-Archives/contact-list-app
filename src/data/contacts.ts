import { uniqueId } from "lodash/fp";
import { IContactFormData } from "types";

export const contactsData: Record<"variant1", IContactFormData[]> = {
  variant1: [
    {
      id: "7cb1248c-983a-4207-a59f-40bfa1fae564",
      name: "Ali Connors",
      gender: "male",
      mobile: "+9282822223",
      email: "ali.connors@whatsapp.com",
      type: "business",
      imgUrl:
        "https://plus.unsplash.com/premium_photo-1671641797679-3b680a7d2109?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MXx8cHJvZmlsZSUyMHBpY3xlbnwwfHwwfHw%3D&auto=format&fit=crop&w=900&q=60",
      company: "SurveyHeart",
      address: "UK",
    },
    {
      id: "b4fc7a3e-3522-4292-b05d-5edbdffcf3ac",
      name: "Jenny Allen",
      gender: "female",
      mobile: "+39292323",
      type: "personal",
      imgUrl:
        "https://images.unsplash.com/photo-1625621435205-2d09ed661a71?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=987&q=80",
      company: "Societe Generale",
    },
    {
      id: "e90ae941-f295-4adc-bf8a-b1d7127a60e6",
      gender: "male",
      mobile: "+9211111214343",
      type: "business",
      imgUrl:
        "https://images.unsplash.com/photo-1500648767791-00dcc994a43e?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxzZWFyY2h8Mnx8cHJvZmlsZSUyMHBpY3xlbnwwfHwwfHw%3D&auto=format&fit=crop&w=900&q=60",
      name: "Steven Covey",
    },
  ],
};

export const getDefaultCreateContact = (): IContactFormData => ({
  id: uniqueId("contact"),
  gender: "female",
  mobile: "",
  name: "",
  type: "personal",
  address: "",
  company: "",
  email: "",
});
