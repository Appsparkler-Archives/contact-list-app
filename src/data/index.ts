export * from "./contacts";
export * from "./components";
