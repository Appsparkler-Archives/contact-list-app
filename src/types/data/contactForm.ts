export interface IContactForm {
  name: string;
  mobile: string;
  address: string;
  company: string;
  email: string;
  gender: "male" | "female";
  type: "personal" | "business";
}
