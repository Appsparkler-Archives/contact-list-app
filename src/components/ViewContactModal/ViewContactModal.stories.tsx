import type { Meta, StoryObj } from "@storybook/react";
import { ViewContactModal } from "./ViewContactModal";
import { contactsData } from "data";

const meta = {
  title: "Components/View Contact Modal",
  component: ViewContactModal,
  tags: ["components"],
} satisfies Meta<typeof ViewContactModal>;

export default meta;

type Story = StoryObj<typeof meta>;

export const viewContactModal: Story = {
  args: {
    contact: contactsData.variant1[0],
  },
};
