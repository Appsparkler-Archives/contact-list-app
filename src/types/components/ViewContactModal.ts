import { IContactFormData } from "types/data";

export type TViewContactModalFC = React.FC<{ contact: IContactFormData }>;

export type TContactInfoItem = React.FC<{
  info?: string;
  icon: React.ReactElement;
}>;
