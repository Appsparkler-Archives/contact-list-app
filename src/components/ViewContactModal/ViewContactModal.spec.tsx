import { act, fireEvent, render, screen } from "@testing-library/react";
import { ViewContactModal } from "./ViewContactModal";
import { contactsData } from "data";

jest.useFakeTimers();

const mockContact = contactsData.variant1[0];

describe("ViewContactModal - Snapshot Tests", () => {
  it("renders correctly when closed", () => {
    const { container } = render(<ViewContactModal contact={mockContact} />);
    expect(container).toMatchSnapshot();
  });

  it("renders correctly when open", () => {
    const { baseElement } = render(<ViewContactModal contact={mockContact} />);
    clickOnTriggerButton();
    expect(baseElement).toMatchSnapshot();
  });
});

describe("ViewContactModal", () => {
  test("should open modal when read button is clicked", () => {
    render(<ViewContactModal contact={mockContact} />);
    fireEvent.click(screen.getByLabelText("read"));
    expect(screen.getByText("View Contact")).toBeInTheDocument();
  });

  test("should close modal when close button is clicked", () => {
    render(<ViewContactModal contact={mockContact} />);
    fireEvent.click(screen.getByLabelText("read"));
    fireEvent.click(screen.getByLabelText("close"));
    act(() => {
      jest.runAllTimers();
    });
    expect(screen.queryByText("View Contact")).not.toBeInTheDocument();
  });

  test("should display contact info when modal is opened", () => {
    render(<ViewContactModal contact={mockContact} />);
    fireEvent.click(screen.getByLabelText("read"));
    expect(screen.getByText(mockContact.name)).toBeInTheDocument();
    expect(screen.getByText(mockContact.mobile)).toBeInTheDocument();
    expect(screen.getByText(String(mockContact.email))).toBeInTheDocument();
    expect(screen.getByText(String(mockContact.address))).toBeInTheDocument();
    expect(screen.getByText(String(mockContact.company))).toBeInTheDocument();
    expect(screen.getByText(mockContact.type)).toBeInTheDocument();
  });

  test("should display avatar with correct color based on gender", () => {
    render(<ViewContactModal contact={mockContact} />);
    fireEvent.click(screen.getByLabelText("read"));
    const avatar = screen.getByLabelText("gender male");
    expect(avatar).toBeInTheDocument();
    expect(avatar).toHaveStyle(`background-color: #2196f3;`); // blue color for male
  });

  test("should not display contact info item if it is not provided", () => {
    const mockContactWithoutEmail = { ...mockContact, email: undefined };
    render(<ViewContactModal contact={mockContactWithoutEmail} />);
    fireEvent.click(screen.getByLabelText("read"));
    expect(
      screen.queryByText(String(mockContactWithoutEmail.email))
    ).not.toBeInTheDocument();
  });
});

function clickOnTriggerButton() {
  fireEvent.click(screen.getByRole("button"));
}
