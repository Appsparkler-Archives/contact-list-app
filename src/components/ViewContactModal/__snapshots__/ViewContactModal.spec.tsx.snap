// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewContactModal - Snapshot Tests renders correctly when closed 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <button
      aria-label="read"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ChromeReaderModeIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M13 12h7v1.5h-7zm0-2.5h7V11h-7zm0 5h7V16h-7zM21 4H3c-1.1 0-2 .9-2 2v13c0 1.1.9 2 2 2h18c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 15h-9V6h9v13z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`ViewContactModal - Snapshot Tests renders correctly when open 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      class="MuiBox-root css-0"
    >
      <button
        aria-label="read"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChromeReaderModeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13 12h7v1.5h-7zm0-2.5h7V11h-7zm0 5h7V16h-7zM21 4H3c-1.1 0-2 .9-2 2v13c0 1.1.9 2 2 2h18c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 15h-9V6h9v13z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r1:"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullScreen css-m9glnp-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <header
          class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-10ghrmp-MuiPaper-root-MuiAppBar-root"
          style="padding-right: 1024px;"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
          >
            <button
              aria-label="close"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-1d6w9lk-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CloseIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <div
              class="MuiTypography-root MuiTypography-h6 css-6b3s0t-MuiTypography-root"
            >
              View Contact
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit Mui-focusVisible MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit css-1y942vo-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              close
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              >
                <span
                  class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
                  style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
                >
                  <span
                    class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
                  />
                </span>
              </span>
            </button>
          </div>
        </header>
        <div
          class="MuiBox-root css-plbgj3"
        >
          <div
            aria-label="gender male"
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-4vb5ob-MuiAvatar-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
              data-testid="MaleIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9.5 11c1.93 0 3.5 1.57 3.5 3.5S11.43 18 9.5 18 6 16.43 6 14.5 7.57 11 9.5 11zm0-2C6.46 9 4 11.46 4 14.5S6.46 20 9.5 20s5.5-2.46 5.5-5.5c0-1.16-.36-2.23-.97-3.12L18 7.42V10h2V4h-6v2h2.58l-3.97 3.97C11.73 9.36 10.66 9 9.5 9z"
              />
            </svg>
          </div>
          <h5
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
             
            Ali Connors
          </h5>
          <div
            class="MuiBox-root css-149sczi"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="PhoneAndroidIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16 1H8C6.34 1 5 2.34 5 4v16c0 1.66 1.34 3 3 3h8c1.66 0 3-1.34 3-3V4c0-1.66-1.34-3-3-3zm-2 20h-4v-1h4v1zm3.25-3H6.75V4h10.5v14z"
              />
            </svg>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              +9282822223
            </h6>
          </div>
          <div
            class="MuiBox-root css-149sczi"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="EmailIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4-8 5-8-5V6l8 5 8-5v2z"
              />
            </svg>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              ali.connors@whatsapp.com
            </h6>
          </div>
          <div
            class="MuiBox-root css-149sczi"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="HouseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 9.3V4h-3v2.6L12 3 2 12h3v8h5v-6h4v6h5v-8h3l-3-2.7zm-9 .7c0-1.1.9-2 2-2s2 .9 2 2h-4z"
              />
            </svg>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              UK
            </h6>
          </div>
          <div
            class="MuiBox-root css-149sczi"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="BusinessIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 7V3H2v18h20V7H12zM6 19H4v-2h2v2zm0-4H4v-2h2v2zm0-4H4V9h2v2zm0-4H4V5h2v2zm4 12H8v-2h2v2zm0-4H8v-2h2v2zm0-4H8V9h2v2zm0-4H8V5h2v2zm10 12h-8v-2h2v-2h-2v-2h2v-2h-2V9h8v10zm-2-8h-2v2h2v-2zm0 4h-2v2h2v-2z"
              />
            </svg>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              SurveyHeart
            </h6>
          </div>
          <div
            class="MuiBox-root css-149sczi"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CategoryIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m12 2-5.5 9h11z"
              />
              <circle
                cx="17.5"
                cy="17.5"
                r="4.5"
              />
              <path
                d="M3 13.5h8v8H3z"
              />
            </svg>
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              business
            </h6>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
